src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs

```diff
@@ -0,0 +1,68 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.Extensions.Caching.Memory
+{
+    internal partial class CacheEntry
+    {
+        // this type exists just to reduce CacheEntry size by replacing many enum & boolean fields with one of a size of Int32
+        [StructLayout(LayoutKind.Explicit)]
+        private struct CacheEntryState
+        {
+            [FieldOffset(0)]
+            private byte _flags;
+            [FieldOffset(1)]
+            private byte _evictionReason;
+            [FieldOffset(2)]
+            private byte _priority;
+            [FieldOffset(3)]
+            private byte _reserved; // for future use
+
+            internal CacheEntryState(CacheItemPriority priority) : this() => _priority = (byte)priority;
+
+            internal bool IsDisposed
+            {
+                get => ((Flags)_flags & Flags.IsDisposed) != 0;
+                set => SetFlag(Flags.IsDisposed, value);
+            }
+
+            internal bool IsExpired
+            {
+                get => ((Flags)_flags & Flags.IsExpired) != 0;
+                set => SetFlag(Flags.IsExpired, value);
+            }
+
+            internal bool IsValueSet
+            {
+                get => ((Flags)_flags & Flags.IsValueSet) != 0;
+                set => SetFlag(Flags.IsValueSet, value);
+            }
+
+            internal EvictionReason EvictionReason
+            {
+                get => (EvictionReason)_evictionReason;
+                set => _evictionReason = (byte)value;
+            }
+
+            internal CacheItemPriority Priority
+            {
+                get => (CacheItemPriority)_priority;
+                set => _priority = (byte)value;
+            }
+
+            private void SetFlag(Flags option, bool value) => _flags = (byte)(value ? (_flags | (byte)option) : (_flags & ~(byte)option));
```

Comment:
Why don't we need to use `Interlocked.CompareExchange` here? Won't we potentially lose changes if 1 thread is setting `IsExpired` while another thread is setting `IsDisposed`?

Note: we were doing that in the previous code:
https://github.com/dotnet/runtime/blob/5078dd61a81c88dbdd131a77b8148b4dee3803fa/src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.cs#L352-L361

Reply:
Other code reviewers asked why it was needed, I've analyzed the code again and it turned out to be not needed (https://github.com/dotnet/runtime/pull/45962#issuecomment-743333631)

> 1 thread is setting IsExpired while another thread is setting IsDisposed

The unusual thing about this particular `Dispose` is that it's being called to... add the entry to the cache. So nobody can set it as expired at the same time because it has not been added to the cache yet

Reply:
The object is handed out to consumers, right? Could someone set the value and call Dispose on it from 2 different threads?

Reply:
> The object is handed out to consumers, right? Could someone set the value and call Dispose on it from 2 different threads?

In theory yes, but in general we don't have any thread safe `Dispose` methods in BCL? Personally I don't see the value in the additional complexity.

Reply:
I just called it out since there is so much other thread safety concerns here (since this is a cache).

If you are satisfied with it, that works for me.